### PR TITLE
Fix photo set intersection bug

### PR DESCRIPTION
**Pull request by kalman**

### CODE REVIEW

File: samples/js/photos/src/photo-set-iterator.js [hunk 1]
```diff
@@ -17,103 +18,157 @@ import type {
...
+    this._lock = false;
   }
 
-  async next(): Promise<PhotoSetIteratorResult> {
-    if (!this._inner && await this._nextInner()) {
+  async next(): Promise<AsyncIteratorResult<IterValue>> {
+    invariant(!this._lock);
+    this._lock = true;
+    try {
+      return await this._next(); // must await so that finally will execute afterwards
```
Comment (kalman) [inline samples/js/photos/src/photo-set-iterator.js]: These changes are actually just a sanity check which is triggering a bug in the photo grid, triggers when you use the date slider. I'll fix that next.

File: samples/js/photos/src/photo-set-iterator.js [hunk 2]
```diff
@@ -17,103 +18,157 @@ import type {
...
+    if (this._done) {
       return {done: true};
     }
-    let next = await notNull(this._inner).next();
-    if (next.done) {
-      if (await this._nextInner()) {
+
+    if (!this._outerIter) {
+      // First call to next().
+      // The outer iterator starts at startDate, which may be -Infinity. If outer iteration started
```
Comment (kalman) [inline samples/js/photos/src/photo-set-iterator.js]: This was the bug.

File: samples/js/photos/src/photo-set-iterator.js [hunk 3]
```diff
@@ -17,103 +18,157 @@ import type {
...
       for (let i = 1; i < values.length; i++) {
         if (!equals(values[i][1], values[0][1])) {
           didIntersect = false;
-          if (!isIterationOrder(values[i][0], values[i][1],
-                                values[last][0], values[last][1])) {
-            last = i;
-          }
+        }
+
+        if (!isIterationOrder(values[i], values[leastRecentIdx])) {
```
Comment (kalman) [inline samples/js/photos/src/photo-set-iterator.js]: This was an accident, I'll change it back.

**Comment by aboodman** [comment]
lgtm

**Pull request merged**